src/v/resource_mgmt/cpu_profiler.cc

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "resource_mgmt/cpu_profiler.h"
+
+#include "random/generators.h"
+#include "resource_mgmt/logger.h"
+#include "ssx/future-util.h"
+#include "ssx/sformat.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/internal/cpu_profiler.hh>
+#include <seastar/core/reactor.hh>
+
+#include <iterator>
+
+namespace resources {
+
+cpu_profiler::cpu_profiler(
+  config::binding<bool>&& enabled,
+  config::binding<std::chrono::milliseconds>&& sample_rate)
+  : _query_timer([this] { return poll_samples(); })
+  , _enabled(std::move(enabled))
+  , _sample_rate(std::move(sample_rate)) {
+    _results_buffer.reserve(seastar::max_number_of_traces);
+    _enabled.watch([this] { on_enabled_change(); });
+    _sample_rate.watch([this] { on_sample_rate_change(); });
+}
+
+ss::future<> cpu_profiler::start() {
+    seastar::engine().set_cpu_profiler_period(_sample_rate());
+    seastar::engine().set_cpu_profiler_enabled(_enabled());
+
+    if (_enabled()) {
+        // Arm the timer to fire whenever the profiler collects
+        // the maximum number of traces it can retain.
+        _query_timer.arm_periodic(ss::max_number_of_traces * _sample_rate());
+    }
+
+    co_return;
+}
+
+ss::future<> cpu_profiler::stop() {
+    _query_timer.cancel();
+    co_await _gate.close();
+}
+
+void cpu_profiler::poll_samples() {
+    ssx::background = ssx::spawn_with_gate_then(
+      _gate, [this] { do_poll_samples(); });
```

Comment:
if your intention is to run do_poll_samples in the background under the gate, this isn't doing that. i think it is running do_poll_samples synchronously with the caller of poll_samples.

Reply:
I originally wasn't sure if I would been asynchronicity in this service so I added the gate in case I did. However, now that it's implemented I know it's not needed. So I'll be removing the gate.

Reply:
Decided to keep the gate in for now. Though I have removed it from this section as it's not needed.

Where I believe it's needed is the `results()` method since I want to ensure all calls to it have completed before the service is stopped.